pkg/server/server.go

```diff
@@ -369,25 +369,23 @@ func (s *ProxyServer) Proxy(stream client.ProxyService_ProxyServer) error {
 	klog.V(5).InfoS("Proxy request from client", "userAgent", userAgent, "serverID", s.serverID)
 
 	recvCh := make(chan *client.Packet, xfrChannelSize)
-	stopCh := make(chan error)
+	stopCh := make(chan error, 1)
 
 	labels := runpprof.Labels(
 		"serverCount", strconv.Itoa(s.serverCount),
 		"userAgent", strings.Join(userAgent, ", "),
 	)
-	go runpprof.Do(context.Background(), labels, func(context.Context) { s.serveRecvFrontend(stream, recvCh) })
-
-	defer func() {
-		close(recvCh)
-	}()
-
 	// Start goroutine to receive packets from frontend and push to recvCh
 	go runpprof.Do(context.Background(), labels, func(context.Context) { s.readFrontendToChannel(stream, userAgent, recvCh, stopCh) })
 
+	s.serveRecvFrontend(stream, recvCh)
```

Comment:
serveRecvFrontend is a blocking function. Before we would return from Proxy() immediately if readFrontendToChannel stream was closed or hit an error. Now we need the full receive channel to be emptied, the channel closed and then the clean up be completed. I'm not saying this is wrong but it is a significant change. Would like to be sure this is actually a safe change.

Reply:
I agree it's significant, but think we should move forward, at least in master branch. In a WIP change, I'm doing something similar anyway (block until receive channel is done processing, then clean any lingering connections).

The main risk is `serveRecvFrontend` failing to return; if that happens we will have good visibility via the metric that will never decrement (`metrics.Metrics.ConnectionDec`). I have not seen this metric have any leak behavior, for recent release tags, so if that changes we will know this likely caused it.